pkg/scheduler/framework/plugins/nodevolumelimits/csi.go

```diff
@@ -77,8 +81,93 @@ func (pl *CSILimits) Name() string {
+	}
+}
+
+func (pl *CSILimits) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	deletedPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		logger.V(5).Error(err, "Unexpected objects in isSchedulableAfterPodDeleted", "oldObj", oldObj)
+		return framework.Queue, fmt.Errorf("unexpected objects in isSchedulableAfterPodDeleted: %w", err)
+	}
+
+	if deletedPod.Namespace != pod.Namespace {
+		return framework.QueueSkip, nil
+	}
+
+	if len(deletedPod.Spec.Volumes) == 0 {
+		return framework.QueueSkip, nil
+	}
+
+	csiNode, err := pl.csiNodeLister.Get(deletedPod.Spec.NodeName)
+	if apierrors.IsNotFound(err) {
+		logger.V(5).Info("csiNodeLister is not found", "nodeName", deletedPod.Spec.NodeName)
+		return framework.QueueSkip, nil
+	} else if err != nil {
+		logger.V(5).Error(err, "Could not get a CSINode object", "nodeName", deletedPod.Spec.NodeName)
+		return framework.Queue, fmt.Errorf("could not get a CSINode object: %w", err)
+	}
+
+	deletedPodDrivers := sets.Set[string]{}
+	for _, vol := range deletedPod.Spec.Volumes {
+		pvcName := ""
+		switch {
+		case vol.PersistentVolumeClaim != nil:
+			pvcName = vol.PersistentVolumeClaim.ClaimName
+		case vol.Ephemeral != nil:
+			pvcName = ephemeral.VolumeClaimName(pod, &vol)
+		default:
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(deletedPod.Namespace).Get(pvcName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.V(5).Info("A PVC is not found", "namespace", pod.Namespace, "pvc name", pvcName)
+				return framework.QueueSkip, nil
+			}
+			logger.V(5).Error(err, "Unable to look up a PVC info", "namespace", pod.Namespace, "pvc name", pvcName)
+			return framework.Queue, fmt.Errorf("unable to look up a PVC info: %w", err)
+
+		}
+
+		driverName, _ := pl.getCSIDriverInfo(logger, csiNode, pvc)
+		deletedPodDrivers.Insert(driverName)
+	}
+
+	for _, vol := range pod.Spec.Volumes {
+		pvcName := ""
+		switch {
+		case vol.PersistentVolumeClaim != nil:
+			pvcName = vol.PersistentVolumeClaim.ClaimName
+		case vol.Ephemeral != nil:
+			pvcName = ephemeral.VolumeClaimName(pod, &vol)
+		default:
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(pvcName)
+		if apierrors.IsNotFound(err) {
+			logger.V(5).Info("A PVC info is not found", "namespace", pod.Namespace, "pvc name", pvcName)
+			return framework.QueueSkip, nil
+		} else if err != nil {
+			logger.V(5).Error(err, "Unable to look up a PVC info", "namespace", pod.Namespace, "pvc name", pvcName)
+			return framework.Queue, fmt.Errorf("unable to look up a PVC info: %w", err)
+		}
+
+		driverName, _ := pl.getCSIDriverInfo(logger, csiNode, pvc)
+
+		for deletedPodDriver := range deletedPodDrivers {
+			if driverName == deletedPodDriver {
+				// This event may make Pod schedulable because the deleted Pod used the same CSI driver as the new Pod.
+				return framework.Queue, nil
+			}
+		}
 	}
```

Comment:
can we replace with `deletedPodDrivers.Has(driverName)`

Reply:
0ed7f7f1f74